shell/platform/android/io/flutter/embedding/FlutterEngine.java

```diff
@@ -0,0 +1,262 @@
+
+    attach();
+    // TODO(mattcarroll): FlutterRenderer is temporally coupled to attach(). Remove that coupling if possible.
+    this.renderer = new FlutterRenderer(flutterJNI, nativeObjectReference);
+  }
+
+  private void attach() {
+    // TODO(mattcarroll): what impact does "isBackgroundView' have?
+    nativeObjectReference = flutterJNI.nativeAttach(this, isBackgroundView);
+
+    if (!isAttached()) {
+      throw new RuntimeException("FlutterEngine failed to attach to its native Object reference.");
+    }
+  }
+
+  private void assertAttached() {
+    if (!isAttached()) throw new AssertionError("Platform view is not attached");
+  }
+
+  @SuppressWarnings("BooleanMethodIsAlwaysInverted")
+  private boolean isAttached() {
+    return nativeObjectReference != 0;
+  }
+
+  public void detach() {
+    pluginRegistry.detach();
+    // TODO(mattcarroll): why do we have a nativeDetach() method? can we get rid of this?
+    flutterJNI.nativeDetach(nativeObjectReference);
+  }
+
+  public void destroy() {
+    pluginRegistry.destroy();
+    flutterJNI.nativeDestroy(nativeObjectReference);
+    nativeObjectReference = 0;
+    applicationIsRunning = false;
+  }
+
+  public void runFromBundle(FlutterRunArguments args) {
+    if (args.bundlePath == null) {
+      throw new AssertionError("A bundlePath must be specified");
+    } else if (args.entrypoint == null) {
+      throw new AssertionError("An entrypoint must be specified");
+    }
+    runFromBundleInternal(args.bundlePath, args.entrypoint, args.libraryPath, args.defaultPath);
+  }
+
+  private void runFromBundleInternal(
+      String bundlePath,
+      String entrypoint,
+      String libraryPath,
+      String defaultPath
+  ) {
+    assertAttached();
+
+    if (applicationIsRunning) {
+      throw new AssertionError("This Flutter engine instance is already running an application");
+    }
+
+    flutterJNI.nativeRunBundleAndSnapshotFromLibrary(nativeObjectReference, bundlePath,
+        defaultPath, entrypoint, libraryPath, resources.getAssets());
+
+    applicationIsRunning = true;
+  }
+
+  public boolean isApplicationRunning() {
+    return applicationIsRunning;
+  }
+
+  // TODO(mattcarroll): what does this callback actually represent?
+  // Called by native to notify when the engine is restarted (cold reload).
+  @SuppressWarnings("unused")
+  private void onPreEngineRestart() {
+    if (pluginRegistry == null)
+      return;
+    pluginRegistry.onPreEngineRestart();
+  }
+
+  public FlutterRenderer getRenderer() {
+    return renderer;
+  }
+
+  public FlutterPluginRegistry getPluginRegistry() {
+    return pluginRegistry;
+  }
+
+  // TODO(mattcarroll): Is this method really all about plugins? or does it have other implications?
+  public void attachViewAndActivity(Activity activity) {
+    pluginRegistry.attach(this, activity);
+  }
+
+  // TODO(mattcarroll): This method says it's unused. Is that true? Do we need it?
+  public String getObservatoryUri() {
```

Comment:
if you keep it, please make sure you use the term "Url" rathen than the obsolete term "Uri" in APIs unless there's a strong reason not to (e.g. because the type is Uri, as sometimes happens in Dart).

Reply:
Changed the public Java API to "Url", leaving the JNI signature to avoid breaking the existing embedding.